Source/SpatialGDK/Private/SpatialInterop.cpp

```diff
@@ -860,16 +867,148 @@ bool USpatialInterop::IsSingletonClass(UClass* Class)
 
 	if (It != View->Entities.end())
 	{
 		improbable::unreal::GlobalStateManagerData* GSM = It->second.Get<improbable::unreal::GlobalStateManager>().data();
-		return &(GSM->singleton_name_to_entity_id());
+		return GSM;
 	}
 
 	return nullptr;
 }
+
+NameToEntityIdMap* USpatialInterop::GetSingletonNameToEntityId() const
+{
+	improbable::unreal::GlobalStateManagerData* GSM = GetGlobalStateManagerData();
+
+	if (GSM == nullptr)
+	{
+		return nullptr;
+	}
+
+	return &(GSM->singleton_name_to_entity_id());
+}
+
+EntityIdToPathMap* USpatialInterop::GetEntityIdToReplicatedStablyNamedPath() const
+{
+	improbable::unreal::GlobalStateManagerData* GSM = GetGlobalStateManagerData();
+
+	if (GSM == nullptr)
+	{
+		return nullptr;
+	}
+
+	return &(GSM->entity_id_to_replicated_stably_named_path());
+}
+
+void USpatialInterop::ReserveReplicatedStablyNamedActorChannel(USpatialActorChannel* Channel)
+{
+	ReplicatedStablyNamedActorChannelQueue.Add(Channel);
+
+	if (CanSpawnReplicatedStablyNamedActors())
+	{
+		ReserveReplicatedStablyNamedActors();
+	}
+}
+
+void USpatialInterop::UnreserveReplicatedStablyNamedActor(AActor* Actor)
+{
+	// Slightly inefficient linear search, could use a hashmap
+	USpatialActorChannel* Channel = nullptr;
+	for (auto& It : ReplicatedStablyNamedActorChannelQueue)
+	{
+		if (It->Actor == Actor)
+		{
+			Channel = It;
+			break;
+		}
+	}
+
+	if (Channel != nullptr)
+	{
+		ReplicatedStablyNamedActorChannelQueue.Remove(Channel);
+	}
+}
+
+void USpatialInterop::AddReplicatedStablyNamedActorToGSM(const FEntityId& EntityId, AActor* Actor)
+{
+	EntityIdToPathMap& EntityIdToReplicatedStablyNamedPathMap = *GetEntityIdToReplicatedStablyNamedPath();
+	// If the map already has the entity id, meaning the actor was already spawned, return early
+	if (EntityIdToReplicatedStablyNamedPathMap.count(EntityId.ToSpatialEntityId()) != 0)
+	{
+		return;
+	}
+
+	std::string Path = TCHAR_TO_UTF8(*Actor->GetPathName(Actor->GetWorld()));
+	EntityIdToReplicatedStablyNamedPathMap.emplace(EntityId.ToSpatialEntityId(), Path);
+
+	improbable::unreal::GlobalStateManager::Update Update;
+	Update.set_entity_id_to_replicated_stably_named_path(EntityIdToReplicatedStablyNamedPathMap);
+
+	TSharedPtr<worker::Connection> Connection = SpatialOSInstance->GetConnection().Pin();
+	Connection->SendComponentUpdate<improbable::unreal::GlobalStateManager>(worker::EntityId((long)SpatialConstants::GLOBAL_STATE_MANAGER), Update);
+}
+
+void USpatialInterop::ReserveReplicatedStablyNamedActors()
+{
+	bCanSpawnReplicatedStablyNamedActors = true;
+	EntityIdToPathMap& EntityIdToReplicatedStablyNamedPath = *GetEntityIdToReplicatedStablyNamedPath();
+
+	for (USpatialActorChannel* Channel : ReplicatedStablyNamedActorChannelQueue)
+	{
+		std::string Path = TCHAR_TO_UTF8(*Channel->Actor->GetPathName(Channel->Actor->GetWorld()));
+
+		// Inefficient, check if we have already spawned this actor
+		bool bFound = false;
+		for (const auto& Pair : EntityIdToReplicatedStablyNamedPath)
```

Comment:
Would it be better if we reversed how this map is stored? Path : id instead? Would make these functions a lot cleaner.

Reply:
Hmm, yes, that would probably make sense, will do!